librz/bin/pdb/pdb.c

```diff
@@ -1,1433 +1,567 @@
-		}
-		rz_buf_seek(pdb->buf, (st64)root_index_pages[i] * (st64)page_size,
-			RZ_BUF_SET);
-		rz_buf_read(pdb->buf, p_tmp, page_size);
-		p_tmp = (char *)p_tmp + page_size;
-	}
-	root_page_list = (int *)calloc(sizeof(int), num_root_pages);
-	if (!root_page_list) {
-		eprintf("Error: memory allocation of root page.\n");
-		goto error;
-	}
-
-	p_tmp = root_page_data;
-	for (i = 0; i < num_root_pages; i++) {
-		root_page_list[i] = *((int *)p_tmp);
-		p_tmp = (int *)p_tmp + 1;
-	}
-
-	pdb->pdb_streams2 = NULL;
-	if (!init_pdb7_root_stream(pdb, root_page_list, num_root_pages,
-		    ePDB_STREAM_ROOT, root_size, page_size)) {
-		eprintf("Could not initialize root stream.\n");
-		goto error;
-	}
-	if (!pdb_read_root(pdb)) {
-		eprintf("PDB root was not initialized.\n");
-		goto error;
-	}
+static char *pdb_type_as_string_json(const RzTypeDB *db, const RzPdb *pdb, const RzList *types) {
+	rz_return_val_if_fail(db && pdb && types, NULL);
 
-	RZ_FREE(root_page_list);
-	RZ_FREE(root_page_data);
-	RZ_FREE(root_index_pages);
-	return true;
-error:
-	RZ_FREE(root_page_list);
-	RZ_FREE(root_page_data);
-	RZ_FREE(root_index_pages);
-	return false;
-}
-
-static void finish_pdb_parse(RzPdb *pdb) {
-	RZ_PDB7_ROOT_STREAM *p = pdb->root_stream;
 	RzListIter *it;
-	SPage *page = 0;
-
-	if (!p) {
-		return;
+	RzBaseType *type;
+	PJ *pj = pj_new();
+	if (!pj) {
+		return NULL;
 	}
-	rz_list_foreach (p->streams_list, it, page) {
-		free(page->stream_pages);
-		page->stream_pages = 0;
-		free(page);
-		page = 0;
-	}
-	rz_list_free(p->streams_list);
-	p->streams_list = 0;
-	free(p);
-	p = 0;
-
-	// TODO: maybe create some kind of destructor?
-	SPDBInfoStream *pdb_info_stream = 0;
-	STpiStream *tpi_stream = 0;
-	SDbiStream *dbi_stream = 0;
-	SStreamParseFunc *stream_parse_func;
-	RZ_PDB_STREAM *pdb_stream = 0;
-	int i = 0;
-	/* rz_list_free should be enough, all the items in a list should be freeable using a generic destructor
-   hacking up things like that may only produce problems. so it is better to not assume that a specific
-   element in a list is of a specific type and just store this info in the type struct or so.
-*/
-	// XXX: this loop is a mess and leaks memory.
-	it = rz_list_iterator(pdb->pdb_streams);
-	while (rz_list_iter_next(it)) {
-		switch (i) {
-		case 1:
-			pdb_info_stream = (SPDBInfoStream *)rz_list_iter_get(it);
-			free_pdb_stream(pdb_info_stream);
-			free(pdb_info_stream);
-			break;
-		case 2:
-			tpi_stream = (STpiStream *)rz_list_iter_get(it);
-			tpi_stream->free_(tpi_stream);
-			free_pdb_stream(tpi_stream);
-			free(tpi_stream);
+	pj_o(pj);
+	rz_list_foreach (types, it, type) {
+		pj_o(pj);
+		switch (type->kind) {
+		case RZ_BASE_TYPE_KIND_STRUCT: {
+			pj_ks(pj, "type", "structure");
+			pj_ks(pj, "name", type->name);
+			pj_kn(pj, "size", type->size);
+			pj_ka(pj, "members");
+			RzTypeStructMember *memb;
+			rz_vector_foreach(&type->struct_data.members, memb) {
+				pj_ks(pj, "member_type", rz_type_as_string(db, memb->type));
```

Comment:
It will leak the pointer. Please don't forget to free it after.

Reply:
This list shares `RzBaseType` with `RzTypeDB`, so we don't need to free them.

Reply:
What list? I meant the `char *` that is returned from `rz_type_as_string()`.

Reply:
Ah, if so, it will leak